### PR TITLE
Feature/CardInputView

### PR DESCRIPTION
**Pull request by bilgehankalkan**
CardInputView implementation. Custom Credit/Debit card input view that supports autofill.

Readme is not included.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
* card-input module created.
* Sample activity added.

## How Has This Been Tested?
Tested on physical device with Android 9.0.

## Checklist:
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
+        context,
+        attrs,
+        defStyleAttr
+    )
+
+    init {
+        if (isInEditMode) View.inflate(context, R.layout.view_card_input, this)
+
```
Comment (selimtoksal) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]: empty line

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            cardInformation = cardInformation.copy(expiryYear = value)
+            field = value
+        }
+    val isDividerVisible = cardBankLogoUrl != null
+
+    fun getCardNumberBackground(context: Context): Drawable? = if (cardNumberValid) {
+        getValidBackgroundDrawable(context)
+    } else {
+        getErrorBackgroundDrawable(context)
+    }
```
Comment (selimtoksal) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt]:
```suggestion
    fun getCardNumberBackground(context: Context): Drawable? =
     if (cardNumberValid) {
        getValidBackgroundDrawable(context)
    } else {
        getErrorBackgroundDrawable(context)
    }
```

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    fun getCvvBackground(context: Context): Drawable? = if (cvvValid) {
+        getValidBackgroundDrawable(context)
+    } else {
+        getErrorBackgroundDrawable(context)
+    }
```
Comment (selimtoksal) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt]:
```suggestion
    fun getExpiryMonthBackground(context: Context): Drawable? = 
    if (expiryMonthValid) {
        getValidBackgroundDrawable(context)
    } else {
        getErrorBackgroundDrawable(context)
    }

    fun getExpiryYearBackground(context: Context): Drawable? = 
    if (expiryYearValid) {
        getValidBackgroundDrawable(context)
    } else {
        getErrorBackgroundDrawable(context)
    }

    fun getCvvBackground(context: Context): Drawable? = 
    if (cvvValid) {
        getValidBackgroundDrawable(context)
    } else {
        getErrorBackgroundDrawable(context)
    }
```

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    fun validateEndGet(): CardInformation? =
+        if (validate()) binding.viewState?.cardInformation else null
+
+    /**
+     *
+     * Resets [CardInputViewState] data including [CardInputViewState.cardInformation] and
+     * validity data.
+     */
+    fun reset() {
+        binding.viewState = binding.viewState?.reset()
```
Comment (bsobe) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]: Generating a new ViewState without any copying process is not the responsibility of view state object.
I mean we need a secondary constructor to generate an empty view state.

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     *
+     * Validates all card fields and if fields are not valid,
+     * it will immediately sets invalid field backgrounds with [CardInputViewState.errorBackgroundColor].
+     *
+     * @see validate
+     *
+     * @return [CardInformation] if all fields are valid, if not returns null.
+     */
+    fun validateEndGet(): CardInformation? =
```
Comment (bsobe) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]:
```suggestion
    fun validateAndGet(): CardInformation? =
```
Typo?

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    private fun setUpView() {
+        with(binding) {
+            with(editTextCardNumber) {
+                setOnEditorActionListener { _, actionId, _ ->
+                    if (actionId == EditorInfo.IME_ACTION_NEXT && viewState?.validationEnabled == true) {
+                        val isValid = validator.isCardNumberValid(text?.toString())
+                        setCardNumberValidity(isValid)
+                        onCardNumberComplete?.invoke(isValid)
+                    }
+                    actionId == EditorInfo.IME_ACTION_NEXT
```
Comment (bsobe) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]: This line should be remove

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            }
+            with(editTextCvv) {
+                setOnEditorActionListener { _, actionId, _ ->
+                    if (actionId == EditorInfo.IME_ACTION_DONE && viewState?.validationEnabled == true) {
+                        val isValid = validator.isCvvValid(text?.toString())
+                        setCardCvvValidity(isValid)
+                        onCvvComplete?.invoke(isValid)
+                        if (isValid) editTextCvv.hideKeyboard()
+                    }
+                    actionId == EditorInfo.IME_ACTION_DONE
```
Comment (bsobe) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]: ++ this line should also be removed

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+data class CardInputViewState(
+    val cardNumberTitle: String,
+    val expiryTitle: String,
+    val cvvTitle: String,
+    private val expiryMonthTitle: String,
+    private val expiryYearTitle: String,
+    val validationEnabled: Boolean = false,
+    val showCvvInfoButton: Boolean = true,
+    @DrawableRes private val inputBackground: Int = R.drawable.shape_card_input_field_background,
```
Comment (bsobe) [inline libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputViewState.kt]:
```suggestion
    @DrawableRes private val inputBackgroundResouce: Int = R.drawable.shape_card_input_field_background,
```
Please add the "resource" end of an object name to show what the object is. Is it a drawable resource or drawable?

File: libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt (hunk 4 above)
Comment (bilgehankalkan) [reply libraries/card-input/src/main/java/com/trendyol/cardinput/CardInputView.kt]: In `reset()`, my goal is keep the initial configurations for `CardInputView`; textColor, backgrounds etc. I tried to achive configuration save without storing it in another state holder or in builder.

**Pull request merged**